ouroboros-network/src/Ouroboros/Network/PeerSelection/LedgerPeers.hs

```diff
@@ -0,0 +1,188 @@
+newtype LedgerPeersConsensusInterface m = LedgerPeersConsensusInterface {
+      lpGetPeers :: SlotNo -> STM m (Maybe [(PoolStake, NonEmpty RelayAddress)])
+    }
+
+data TraceLedgerPeers =
+      PickedPeer !RelayAddress !AckPoolStake ! PoolStake
+    | PickedPeers !Word16 ![RelayAddress]
+    | FetchingNewLedgerState !Int
+
+
+instance Show TraceLedgerPeers where
+    show (PickedPeer addr ackStake stake) =
+        printf "PickedPeer %s ack stake %s ( %.04f) relative stake %s ( %.04f )"
+            (show addr)
+            (show ackStake) (fromRational ackStake :: Double)
+            (show stake) (fromRational stake :: Double)
+    show (PickedPeers n peers) =
+        printf "PickedPeers %d %s" n (show peers)
+    show (FetchingNewLedgerState cnt) =
+        printf "Fetching new ledgerstate, %d registered pools"
+            cnt
+
+
+data RelayAddress = RelayAddressDomain DomainAddress
+                  | RelayAddressAddr IP.IP Socket.PortNumber
+                  deriving (Show, Eq, Ord)
+
+-- | The relative stake of the stakepool. A value in the [0, 1] range.
+type PoolStake = Rational
+
+-- | The relative stake of the stakepool and all preceeding pools. A value in the range [0, 1].
+type AckPoolStake = Rational
+
+-- | Convert a list of pools with stake to a Map keyed on the ackumulated stake.
+--
+ackPoolStake :: [(PoolStake, NonEmpty RelayAddress)]
+             -> Map Rational (PoolStake, NonEmpty RelayAddress)
+ackPoolStake pl =
+    let pl' = reRelativeStake pl
+        ackList = snd $ foldl' (\(as, ps) (s, rs) -> (as + s, (as + s, (s, rs)):ps)) (0, []) pl' in
+    Map.fromList ackList
+
+-- | Not all stake pools have valid/usable relay information. This means that we need to
+-- recalculate the relative stake for each pool.
+reRelativeStake :: [(PoolStake, NonEmpty RelayAddress)]
+                -> [(PoolStake, NonEmpty RelayAddress)]
+reRelativeStake pl =
+    let total = sum $ map fst pl in
+    map (\(s, rls) -> (s / total, rls)) pl
+
+-- try to pick n random peers
+pickPeers :: forall m. Monad m
+          => StdGen
+          -> Tracer m TraceLedgerPeers
+          -> Map Rational (PoolStake, NonEmpty RelayAddress)
+          -> Word16
+          -> m (StdGen, [RelayAddress])
+pickPeers inRng _ pools _ | Map.null pools = return (inRng, [])
+pickPeers inRng tracer pools cnt = go inRng cnt []
+  where
+    go :: StdGen -> Word16 -> [RelayAddress] -> m (StdGen, [RelayAddress])
+    go rng 0 picked = return (rng, picked)
+    go rng n picked =
+        let (r :: Word64, rng') = random rng
+            d = maxBound :: Word64
+            x = fromIntegral r % fromIntegral d
+            !pick_m = Map.lookupGE x pools in
+        case pick_m of
+             Nothing -> go rng' (n - 1) picked -- XXX We failed pick a peer. Shouldn't this be an error?
+             Just (ackStake, (stake, relays)) -> do
+                 let (ix, rng'') = randomR (0, NonEmpty.length relays - 1) rng'
+                     relay = relays NonEmpty.!! ix
```

Comment:
We might pick the same `relay` multiple times, is that ok?

Reply:
I think that should be fine, there could also be overlap due to different domain names resolving to the same address. We may also pick a peer that the governor already knows about, or the peer may be unreachable.
The governor asks LedgerPeers for some fresh IP addresses, if some of them turn out to be unsuitable it can simply ask again.

Reply:
That's ok, thanks for clarification.